### ISSUE TITLE
Grafana 11: slow load of dashboard

### ISSUE THREAD

**Issue by frenkye** [grafana/grafana#88136]
### What happened?

We have dashboard for system metrics / node_exporter. There is variable filter by `node_boot_time_seconds`, this query in prometheus over 24h is resulted in 0,5s on our side ~800 servers.

After upgrading grafana to 11.0.0 is stucked on 2 queries viz screenshot and the dashboard load hang until those 2 queries returned data
<img width="1032" alt="Screenshot 2024-05-21 at 16 06 54" src="https://github.com/grafana/grafana/assets/13994937/2785c4fa-156f-4902-8d01-7f0ff70c4cdb">

If i downgrade back to 10.4.3 the dashboard load <1s vs ~7s on 11.0.0

Load time is the same with scenes feature enable and without it.

### What did you expect to happen?

Load time will be the same.

### Did this work before?

yes

### How do we reproduce it?

1. Upgrade to 11.0.0
2. Load dashboard

### Is the bug inside a dashboard panel?

whole dashboard

### Environment (with versions)?

Grafana: 11.0.0
OS: Debian 10
Browser: Firefox latest

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

### Datasource(s)?

prometheus

**Comment by tonypowa** [comment]
Hi @frenkye 

Thank you for reporting this issue

Can you share the exact queries or a sanitized version that is causing the slow dashboard load times? This may help identify if a specific query is at fault.

Are there any console errors in the browser developer tools when the dashboard is loading? If yes, please provide the error messages or screenshots.

Tagging both @grafana/dashboards-squad  and @grafana/observability-metrics  squads

**Comment by ivanortegaalba** [comment]
Hi @frenkye, the HAR file would help a lot to see more about where the time is going :)

**Comment by frenkye** [comment]
@tonypowa @ivanortegaalba 

Thank you for replies, all files provided has been sanitized. There should be still clear distinct between grafana and server urls.

1. there are no console erorrs leading to slow quries, there is 1-2s later one, but thats something about annotation list. Happen in both 10.4.3 and 11.0.0, so this is not related.
2. Created HAR files from 10.4.3 and 11.0.0 
[grafana-10-4-3-har.json](https://github.com/user-attachments/files/15910271/grafana-10-4-3-har.json)
[grafana-11-0-0-har.json](https://github.com/user-attachments/files/15910272/grafana-11-0-0-har.json)
3. Copy the one of slow queries as cUrl request, when i call it from shell on server it return response after 5-6s as well.
[grafana-slow-query-curl.txt](https://github.com/user-attachments/files/15910293/grafana-slow-query-curl.txt)
4. If it help, this is our dashboard 
[System metrics-1718867201099.json](https://github.com/user-attachments/files/15910298/System.metrics-1718867201099.json)

**Comment by frenkye** [comment]
Just tested 11.1 and it is the same issue.

Maybe little more info to this, seems to Grafana does query for variables based on time frame of dashboard.
- 24h load ~6s (2x 3s in har)
- 5m load ~2s (2x 1x in har)

It is not on our end in prometheus as he served 24h data between 700ms to 1200ms and the 10.4.3 works ok.
<img width="1667" alt="Screenshot 2024-07-08 at 9 19 40" src="https://github.com/grafana/grafana/assets/13994937/1b6f3185-2909-4a1c-9869-ae1714dae5d8">

**Comment by bohandley** [comment]
Hello! The har files you provided we could not use, can you please upload them again? I think there might have been a sanitization tool that did some weird stuff to them.

One thing to note is that between these versions, we moved to default support for the label_values with match[] parameter for Prometheus instances that had not selected a type and version in the Grafana Prometheus data source configuration page.

![Image](https://github.com/grafana/grafana/assets/25674746/192d3e8b-ecaa-4181-be66-183d6d85c083)

So, if you had not selected a type and version before, it would call the series endpoint for the variable of query type label values. In version 11 we don't use the series endpoint. 

Do you have a Prometheus version that does not support the match parameter for the label values endpoint? What is your Prometheus versions? I have a hunch that this could be fixed by selecting type and version in the config. Thank you!

**Comment by frenkye** [comment]
This was a bit helpfull.

We don't have selected version for prometheus in DS definition. When i do that, the behavior copy Grafana 11 and if I deselect version of prometheus in DS it goes back to ~1s load.

We are currently on production have 2.49.1 and preparing to upgrade on 2.50.x, but it will be migration to full new fault redundant stack which i hope will be done end of Summer.

Is there some way how to speed this new behavior up? Because it is unuseable.

**Comment by mmospanenko** [comment]
having the same behavior after updating from v10. Docker, Loki and Prometheus - all requests extremely slow (in tens times slower than before).

**Comment by mmospanenko** [comment]
@gustavoffilho I had wrong configuration after update, also I moved to another s3 provider (own Minio), I didn't migrate old data. There were a lot of information in the logs

**Comment by alphabet5** [comment]
I'm seeing a similar slowdown when upgrading from 10.x->11.

**Comment by frenkye** [comment]
I circled back to this, since it is blocking our upgrade.

Grafana 10.4.3 sent for 90 days w/o prometheus version datasource:
```
POST /api/datasources/uid/000000003/resources/api/v1/series
with payload
--data-raw 'match%5B%5D=node_boot_time_seconds&start=1719666014&end=1727442014'

When tested as curl copy from shell
real    0m0.804s
user 0m0.018s
sys     0m0.003s

From nginx log file:
"POST /api/v1/series HTTP/1.1" 200 28628 "-" "Grafana/10.4.3" 0.248 661 45138 - "-" 0.248
```

Grafana 11.x or Grafana 10.4.3 w/ prometheus version datasource for 90 days :
```
GET /api/datasources/uid/000000003/resources/api/v1/label/instance/values?match%5B%5D=node_boot_time_seconds&start=1719668423&end=1727444423

real    0m17.945s
user 0m0.011s
sys     0m0.006s

From nginx log file:
"GET /api/v1/label/instance/values?match%5B%5D=node_boot_time_seconds&start=1719668423&end=1727444423 HTTP/1.1" 200 4007 "http://prometheus.mon.overkill.cz/api/v1/label/instance/values?match%5B%5D=node_boot_time_seconds&start=1719668423&end=1727444423" "Grafana/11.2.1" 17.924 799 56898 - "-" 17.924
```

I'am inclined to fact the new query is heavy under optimized on larger setups. Plain query without match
```
time curl -s http://127.0.0.1:9090/api/v1/label/instance/values -o /dev/null

real	0m0.031s
user 0m0.006s
sys	0m0.005s
```
but when we add filter
```
time curl -s 'http://127.0.0.1:9090/api/v1/label/instance/values?match[]=node_boot_time_seconds' -o /dev/null

real	0m17.522s
user 0m0.000s
sys	0m0.009s
```

**Comment by ashleydavies** [comment]
I'm also experiencing incredibly slow user experience with this endpoint. We're changing to series queries with regexes too

**Comment by bohandley** [comment]
For users who would prefer to enable the series endpoint even if their Prometheus version supports label values with match[] param, we have added the following toggle to the Grafana data source Prometheus configuration
https://github.com/grafana/grafana/pull/94443

**Comment by ashleydavies** [comment]
Oh, thanks @bohandley! You posted that just in time to save me refactoring dozens of usages 😄 appreciate the heads up.

I'm curious why some of us seem to experience significant performance regressions from this new endpoint (I have no proxies or similar in-between Grafana and Prometheus), but I'm happy to take the easy W here and unblock users.

**Comment by frenkye** [comment]
Nice 👍  Looking forward to 11.3.2  / 11.4. Thank you very much for working on this!

**Comment by briend** [comment]
Has anyone confirmed this is fixed in `11.3.2` and/or `11.4.0` with the `Series Endpoint` setting?  I could not find the setting yet and it seemed to me to be missing in the releases: https://github.com/grafana/grafana/pull/94443#issuecomment-2546634620.   However, forcing an older version on the Datasource configuration seems to be an adequate work-around at the moment (this prevents the usage of the newer `values` endpoint):

![Image](https://github.com/user-attachments/assets/2f2ff6d3-d073-48d4-b558-db4f642438d3)

I'm not totally sure what the consequence of forcing an old version like this is. . . We're using Prometheus `2.51.2` and Thanos `0.34.1`

**Comment by frenkye** [comment]
Its not in any released version. We have to wait until next release. If you want old behavier you need to be on latest 10.x and have unselected version of Prometheus.

**Comment by briend** [comment]
@frenkye it is confusing because @itsmylife stated here that it was indeed in `11.3`: https://github.com/grafana/grafana/pull/94443#issuecomment-2539226190

but like I posted above, setting Prometheus version to `< 2.14.x` seems to give you the old (faster) behavior (tested on Grafana `11.1.3`)

**Comment by frenkye** [comment]
@briend Ah. Seems somebody missed addition to tag then. 

You are right selecting `< 2.14.x` bring back series endpoint just tested this on 11.4. Time to do some work on dashboards before holidays

**Comment by melroy89** [comment]
Since  Grafana v11.3 I also have _very slow_ dashboard loading times using Influxdb v2.

**Comment by frenkye** [comment]
Hey everyone, it is now officially in 11.5 and it works with prometheus selected on proper version.

Thank you @bohandley again for the fix 🥇

**Comment by melroy89** [comment]
> Since Grafana v11.3.0 I also have _very slow_ dashboard loading times using Influxdb v2 as my primary data-source.
> 
> Downgrading to Grafana v11.2.3 solved /workaround it for now.. But that is of course not ideal.
> 
> **What did you do in Grafana v11.3.0!??** No, it's not due to influxdb, it's Grafana.
> 
> Before (Grafana v11.2.3):
> grafana_before_this_is_version_11.2.3.mp4
> 
> After (Grafana v11.3.0):
> grafana_after_this_is_version_11.3.0.mp4

And is this regression also fixed?

**Comment by frenkye** [comment]
@melroy89 Maybe don't try reopen closed and resolved issue and create a new one with link to this one for refference. Since it is closed and wont get too many attention ...

**Comment by melroy89** [comment]
I didn't reopen anything :)

**Comment by bohandley** [comment]
Hi @melroy89! [I see your comment here on slow dashboard loading with InfluxDB](https://github.com/grafana/grafana/issues/88136#issuecomment-2557911653). While the [original issue](https://github.com/grafana/grafana/issues/88136) is focused on the slow dashboard loading with Prometheus, the InfluxDB regression was not addressed in this issue. 

if you would like to get more info/investigation/a fix on slow dashboard loading with InfluxDB, I would create a new issue and tag the @grafana/oss-big-tent squad. Thank you!
